riemann_client/client.py

```diff
@@ -200,3 +204,143 @@ def send_events(self, events):
+        self.event_counter = 0
+        self.last_flush = time.time()
+        self.timer = None
+
+        # start the timer
+        self.start_timer()
+
+    def connect(self):
+        """Connect the transport if it is not already connected."""
+        if not self.is_connected():
+            self.transport.connect()
+
+    def is_connected(self):
+        """Check whether the transport is connected."""
+        try:
+            # this will throw an exception whenever socket isn't connected
+            self.transport.socket.type
+            return True
+        except (AttributeError, RuntimeError, socket.error):
+            return False
+
+    def event(self, **data):
+        """Enqueues an event, using keyword arguments to create an Event
+
+        >>> client.event(service='riemann-client', state='awesome')
+
+        :param \*\*data: keyword arguments used for :py:func:`create_event`
+        """
+        self.send_events((self.create_event(data),))
+
+    def events(self, *events):
+        """Enqueues multiple events in a single message
+
+        >>> client.events({'service': 'riemann-client', 'state': 'awesome'})
+
+         :param \*events: event dictionaries for :py:func:`create_event`
+         :returns: The response message from Riemann
+        """
+        self.send_events(self.create_event(evd) for evd in events)
+
+    def send_event(self, event):
+        """Enqueues a single event
+
+        :param event: An ``Event`` protocol buffer object
+        """
+        self.send_events((event,))
+
+    def send_events(self, events):
+        """Enqueues multiple events
+
+        :param events: A list or iterable of ``Event`` objects
+        :returns: The response message from Riemann
+        """
+        with self.lock:
+            for event in events:
+                self.queue.events.add().MergeFrom(event)
+                self.event_counter += 1
+                self.check_for_flush()
+
+    def flush(self):
+        """Sends the events in the queue to Riemann in a single protobuf
+        message
+
+        :returns: The response message from Riemann
+        """
+        with self.lock:
+            if not self.is_connected():
+                self.connect()
+            try:
+                response = super(AutoFlushingQueuedClient, self).flush()
+            except socket.error:
+                # log and retry
+                logging.warn("Socket error while flushing."
+                             "Attempting reconnect and retry...")
+                try:
+                    self.transport.disconnect()
+                    self.connect()
+                    response = super(AutoFlushingQueuedClient, self).flush()
+                except:
+                    logging.error("Socket error when flushing"
+                                  "#2. Batch discarded.")
+                    logging.exception()
+                    self.transport.disconnect()
+                    self.clear_queue()
```

Comment:
There's no clearly winning choice here... you either end up with an ever-growing queue while the other end of the socket is down (so you either run out of memory, or try to send massive batches potentially affecting latencies across the entire system), or you lose messages. Eventually you'll lose messages anyway when you run out of memory and local storage capacity, though. 

How about making it tunable via an init parameter like `clear_on_fail`?